### PR TITLE
feat: onComponentTypeCheck for Sweep

### PR DESCRIPTION
**Pull request by ASGAlex**
# Description

Add onComponentTypeCheck functionality to Sweep.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have followed the [Contributor Guide] when preparing my PR.
- [ ] I have updated/added tests for ALL new/updated/fixed functionality.
- [ ] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [ ] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

This is a quession for discussion. I pushed not best impllementation but it avoids breaking changes. We should to decide here, what to do.

- [ ] Yes, this PR is a breaking change.
- [ ] No, this PR is not a breaking change.

### CODE REVIEW

File: packages/flame/lib/src/collisions/broadphase/sweep/sweep.dart [hunk 1]
```diff
@@ -27,18 +31,50 @@ class Sweep<T extends Hitbox<T>> extends Broadphase<T> {
...
     }
+    if (_potentialsTmp.isNotEmpty) {
+      for (var i = 0; i < _potentialsTmp.length; i++) {
+        final item0 = _potentialsTmp[i].first;
+        final item1 = _potentialsTmp[i].last;
+        final checkFunction = broadphaseCheck;
+        var checkSuccess = false;
+        if (checkFunction != null &&
+            item0 is ShapeHitbox &&
+            item1 is ShapeHitbox) {
```
Comment (spydon) [inline packages/flame/lib/src/collisions/broadphase/sweep/sweep.dart:59]: We don't want to have type checks in here, everything should be based of T.
Possibly the sweep could be chained with another step before it that has a tighter restriction on its generics.

**Review by spydon — COMMENTED** [review]
Not too fond of this draft unfortunately, it breaks a lot of the generalization of the class. But maybe that logic can be broken out and wrap the sweep instead and then that class could be used by the `StandardCollisionDetection`.

**Comment by ASGAlex** [comment]
> Not too fond of this draft unfortunately, it breaks a lot of the generalization of the class. But maybe that logic can be broken out and wrap the sweep instead and then that class could be used by the `StandardCollisionDetection`.

I'm not happy too. Code looks ugly, it is strictly dependent on PositionComponents and ShapeHitbox, it breaks some tests... Right now i measuring it's performance because I still have questions with these. 

That's why I said that maybe it would be better to leave this unimplemented :-)

**Comment by spydon** [comment]
> > Not too fond of this draft unfortunately, it breaks a lot of the generalization of the class. But maybe that logic can be broken out and wrap the sweep instead and then that class could be used by the `StandardCollisionDetection`.
> 
> I'm not happy too. Code looks ugly, it is strictly dependent on PositionComponents and ShapeHitbox, it breaks some tests... Right now i measuring it's performance because I still have questions with these.
> 
> That's why I said that maybe it would be better to leave this unimplemented :-)

It feels like it should be possible to do this and still follow the generics, since the `GenericCollisionCallbacks` also have this method. But the `ExternalBroadphaseCheck` typedef can't be used since it is bound to `PositionComponent`.

**Comment by spydon** [comment]
I'll close this for now and maybe we can re-do something similar in the future

**Pull request closed without merging**